pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,308 @@
+			Expect(expectedTyped.GroupVersionKind().Kind).
+				To(Equal(expectedKind))
+		})
+
+		It("the returned resource's Group should be the default KubeFed FederatedGroup", func() {
+			Expect(expectedTyped.GroupVersionKind().Group).
+				To(Equal(kubefedopt.DefaultFederatedGroup))
+		})
+
+		It("the returned resource's Version should be the default KubeFed FederatedVersion", func() {
+			Expect(expectedTyped.GroupVersionKind().Version).
+				To(Equal(kubefedopt.DefaultFederatedVersion))
+		})
+
+		It("the returned resource's name should be the same as the input's object", func() {
+			Expect(expectedTyped.GetName()).
+				To(Equal(resourceTyped.GetName()))
+		})
+
+		It("the returned resource's template should be the input's unstructured", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+
+			mapObj, found, err := getTemplateField(expectedTyped)
+			targetResource := &unstructured.Unstructured{}
+			// No need to check error code because it's covered by another test
+			_ = f.scheme.Convert(resourceTyped, targetResource, nil)
+			removeUnwantedFields(targetResource)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj["spec"]).To(BeEquivalentTo(targetResource.Object["spec"]))
+			Expect(mapObj["metadata"]).To(BeEquivalentTo(targetResource.Object["metadata"]))
+		})
+
+		It("the returned resource's placement should be an empty matcher", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			mapObj, found, err := getMatchLabelsField(expectedTyped)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+			Expect(mapObj).To(BeEmpty())
+		})
+
+		It("the returned resource's namespace should be the input's namespace", func() {
+			Expect(expectedTyped.GetNamespace()).
+				To(Equal(resourceTyped.GetNamespace()))
+		})
+	})
+}
+
+// func assertNestedFields(accessor interface{}, expectedObj types.GomegaMatcher, obj map[string]interface{}, fields ...string) { }
+func testDistribute() {
+	var (
+		resourceTyped       *corev1.Pod
+		fedResource, fedPod *unstructured.Unstructured
+		f                   *federator
+		err                 error
+		initObjs            []runtime.Object
+		clusterNames        []string
+		scheme              *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resourceTyped = newPod("test-pod")
+		scheme = runtime.NewScheme()
+	})
+
+	JustBeforeEach(func() {
+		f = newFederatorWithScheme(scheme, initObjs...)
+		err = f.Distribute(resourceTyped, clusterNames...)
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		It("should throw an error", func() {
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			corev1.AddToScheme(scheme)
+			tmpFederator := newFederatorWithScheme(scheme)
```

Comment:
I find it weird that you need another federator here, since it'll also use another client with another cache, but you only use it to create the "federated resource" and then dump it..

IMHO it's confusing and a bit counter-intuitive.

Might I suggest that you make createFederatedResource and independent method that receives the scheme as a parameter, which should help simplify this situation (and also the tests for createFederatedResource would be a bit simpler).
I tested it locally and it's super easy, barely an inconvenience ;)